File: 54.spiral-matrix.java
/*
 * @lc app=leetcode id=54 lang=java
 *
 * [54] Spiral Matrix
 */

// @lc code=start

import java.util.ArrayList;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>(matrix.length * matrix[0].length);

        int rowBegin = 0;
        int rowEnd = matrix.length - 1;
        int colBegin = 0;
        int colEnd = matrix[0].length - 1;

        while (rowBegin <= rowEnd && colBegin <= colEnd) {

            for (int col = colBegin; col <= colEnd; col++) {
                res.add(matrix[rowBegin][col]);
            }
            rowBegin++;

            for (int row = rowBegin; row <= rowEnd; row++) {
                res.add(matrix[row][colEnd]);
            }
            colEnd--;

            if (rowBegin <= rowEnd) {
                for (int col = colEnd; col >= colBegin; col--) {
                    res.add(matrix[rowEnd][col]);
                }
            }
            rowEnd--;

            if (colBegin <= colEnd) {
                for (int row = rowEnd; row >= rowBegin; row--) {
                    res.add(matrix[row][colBegin]);
                }
            }
            colBegin++;

        }

        return res;

    }
}
// @lc code=end

File: 54.spiral-matrix.py
#
# @lc app=leetcode id=54 lang=python3
#
# [54] Spiral Matrix
#

# @lc code=start
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        res = []
        row_begin = 0
        row_end = len(matrix)- 1
        col_begin = 0
        col_end = len(matrix[0]) - 1

        while row_begin <= row_end and col_begin <= col_end:
            for col in range(col_begin, col_end + 1):
                res.append(matrix[row_begin][col])
            row_begin += 1

            for row in range(row_begin, row_end + 1):
                res.append(matrix[row][col_end])
            col_end -= 1

            if row_begin <= row_end:
                for col in range(col_end, col_begin - 1, -1):
                    res.append(matrix[row_end][col])
                row_end -= 1
            
            if col_begin <= col_end:  
                for row in range(row_end, row_begin - 1, -1):
                    res.append(matrix[row][col_begin])
                col_begin += 1
        
        return res
                

        
            

# @lc code=end

